plugins/woocommerce/src/Internal/StockNotifications/Notification.php

```diff
@@ -449,4 +458,29 @@ public function get_product_name() {
 
 		return $product->get_parent_id() ? $product->get_name() : $product->get_title();
 	}
+
+	public function get_email_link( $action ) {
+		$key = wp_generate_password( 20, false );
+		$hashed = wp_fast_hash( $key );
+
+		$this->set_email_link_action_key( $hashed );
+		$this->save();
+
+		return add_query_arg(
+			array(
+				'action' => $action,
+				'key'    => $key,
+				'id'     => $this->get_id(),
+			),
+			get_site_url()
+		);
+	}
+
+	public function verify_email_link_action_key( $key ) {
+		$hash = $this->get_email_link_action_key();
+		if (  empty( $hash ) || empty( $key ) ) {
+			return false;
+		}
+		return wp_verify_fast_hash( $key, $hash );
+	}
```

Comment:
The WordPress's process for storing and verifying password reset links is:
- create a key with `wp_generate_password`
- store the key hashed and expiration time in the database
- email a link with the unhashed key to the user

This process i'm proposing here is the same, minus the expiration time. 
I figure an unsubscribe link should never expire, so we can omit that step.

Reply:
> I figure an unsubscribe link should never expire, so we can omit that step.

Yup! We only need an expiration for the verification process! Based on my previous comment, the expiration check can happen inside the `has_valid_verification_data()` method.